nemo/collections/nlp/models/qa_model.py

```diff
@@ -0,0 +1,218 @@
+        """
+        # init superclass
+        super().__init__()
+        self.bert_model = get_pretrained_lm_model(pretrained_model_name=pretrained_model_name, config_file=config_file)
+        self.hidden_size = self.bert_model.config.hidden_size
+        self.tokenizer = get_tokenizer(pretrained_model_name=pretrained_model_name, tokenizer_name="nemobert")
+        self.classifier = TokenClassifier(
+            hidden_size=self.hidden_size,
+            num_classes=num_classes,
+            num_layers=num_layers,
+            activation=activation,
+            log_softmax=log_softmax,
+            dropout=dropout,
+            use_transformer_init=use_transformer_init,
+        )
+
+        self.loss = SpanningLoss()
+        # This will be set by setup_training_datai
+        self.__train_dl = None
+        # This will be set by setup_validation_data
+        self.__val_dl = None
+        # This will be set by setup_test_data
+        self.__test_dl = None
+        # This will be set by setup_optimization
+        self.__optimizer = None
+        # This will be set by setup_optimization
+        self.__scheduler = None
+
+    @typecheck()
+    def forward(self, input_ids, token_type_ids, attention_mask):
+        """
+        No special modification required for Lightning, define it as you normally would
+        in the `nn.Module` in vanilla PyTorch.
+        """
+        hidden_states = self.bert_model(
+            input_ids=input_ids, token_type_ids=token_type_ids, attention_mask=attention_mask
+        )
+        logits = self.classifier(hidden_states=hidden_states)
+        return logits
+
+    def training_step(self, batch, batch_idx):
+        """
+        Lightning calls this inside the training loop with the data from the training dataloader
+        passed in as `batch`.
+        """
+        # forward pass
+        input_ids, input_type_ids, input_mask, unique_ids, start_positions, end_positions = batch
+        logits = self.forward(input_ids=input_ids, token_type_ids=input_type_ids, attention_mask=input_mask)
+        loss, _, _ = self.loss(logits=logits, start_positions=start_positions, end_positions=end_positions)
+
+        tensorboard_logs = {'train_loss': loss}
+        return {'loss': loss, 'log': tensorboard_logs}
+
+    def validation_step(self, batch, batch_idx):
+        """
+        Lightning calls this inside the validation loop with the data from the validation dataloader
+        passed in as `batch`.
+        """
+        # forward pass
+        input_ids, input_type_ids, input_mask, unique_ids, start_positions, end_positions = batch
+        logits = self.forward(input_ids=input_ids, token_type_ids=input_type_ids, attention_mask=input_mask)
+        loss, start_logits, end_logits = self.loss(
+            logits=logits, start_positions=start_positions, end_positions=end_positions
+        )
+
+        tensorboard_logs = {'val_loss': loss}
+        return {'val_loss': loss, 'log': tensorboard_logs}
```

Comment:
please remove tensorboard_logs from validation_step function. 
log validation loss (and actually other metrics too) after aggregation inside validation_epoch_end function. This way validation loss/metrics are per validation epoch

Reply:
done